cmd/get_client_certificate.go

```diff
@@ -372,15 +391,88 @@ func (o *GetClientCertificateOptions) getExecCredential(ctx context.Context, cer
 			Kind:       "ExecCredential",
 		},
 		Status: &clientauthenticationv1.ExecCredentialStatus{
-			ExpirationTimestamp:   &adminKubeconfigRequest.Status.ExpirationTimestamp,
+			ExpirationTimestamp:   &kubeconfigRequest.expirationTimestamp,
 			ClientCertificateData: string(certificateSet.ClientCertificateData),
 			ClientKeyData:         string(certificateSet.ClientKeyData),
 		},
 	}, nil
 }
 
-func createAdminKubeconfigRequest(ctx context.Context, client rest.Interface, namespace string, shootName string, adminKubeconfigRequest *authenticationv1alpha1.AdminKubeconfigRequest, opts metav1.CreateOptions) (*authenticationv1alpha1.AdminKubeconfigRequest, error) {
-	result := &authenticationv1alpha1.AdminKubeconfigRequest{}
+func (o *GetClientCertificateOptions) createKubeconfigRequest(ctx context.Context) (*kubeconfigRequestStatus, error) {
+	logger := klog.FromContext(ctx)
+
+	switch o.AccessLevel {
+	case accessLevelAdmin:
+		return createAdminKubeconfigRequest(
+			ctx,
+			o.GardenCoreV1Beta1RESTClient,
+			o.ShootRef.Namespace,
+			o.ShootRef.Name,
+			o.KubeconfigExpirationSeconds,
+			metav1.CreateOptions{},
+		)
+	case accessLevelViewer:
+		return createViewerKubeconfigRequest(
+			ctx,
+			o.GardenCoreV1Beta1RESTClient,
+			o.ShootRef.Namespace,
+			o.ShootRef.Name,
+			o.KubeconfigExpirationSeconds,
+			metav1.CreateOptions{},
+		)
+	case accessLevelAuto:
+		kubeconfigRequest, err := createAdminKubeconfigRequest(
+			ctx,
+			o.GardenCoreV1Beta1RESTClient,
+			o.ShootRef.Namespace,
+			o.ShootRef.Name,
+			o.KubeconfigExpirationSeconds,
+			metav1.CreateOptions{},
+		)
+
+		if err == nil {
+			return kubeconfigRequest, nil
+		}
+
+		if !apierrors.IsForbidden(err) {
+			return nil, err
+		}
+
+		logger.Info("No permission to obtain admin kubeconfig. Falling back to obtaining viewer kubeconfig.", "error", err)
+
+		kubeconfigRequest, err = createViewerKubeconfigRequest(
+			ctx,
+			o.GardenCoreV1Beta1RESTClient,
+			o.ShootRef.Namespace,
+			o.ShootRef.Name,
+			o.KubeconfigExpirationSeconds,
+			metav1.CreateOptions{},
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to obtain either admin or viewer kubeconfig: %w", err)
+		}
+
+		return kubeconfigRequest, nil
```

Comment:
In this case I think the nesting makes it more readable. What do you think?
```go
                // Attempt to create admin kubeconfig request
                kubeconfigRequest, err := createAdminKubeconfigRequest(
                    ctx,
                    o.GardenCoreV1Beta1RESTClient,
                    o.ShootRef.Namespace,
                    o.ShootRef.Name,
                    o.KubeconfigExpirationSeconds,
                    metav1.CreateOptions{},
                )
                
                if apierrors.IsForbidden(err) {
                    // Log and attempt to create viewer kubeconfig request
                    logger.Info("No permission to obtain admin kubeconfig. Falling back to obtaining viewer kubeconfig.", "error", err)
                    kubeconfigRequest, err = createViewerKubeconfigRequest(
                        ctx,
                        o.GardenCoreV1Beta1RESTClient,
                        o.ShootRef.Namespace,
                        o.ShootRef.Name,
                        o.KubeconfigExpirationSeconds,
                        metav1.CreateOptions{},
                    )
                    if apierrors.IsForbidden(err) {
                        return nil, fmt.Errorf("no permission to obtain either admin or viewer kubeconfig: %w", err)
                    }
                }
                
                if err != nil {
                    return nil, err
                }
                
                // Return the kubeconfigRequest which might be either admin or viewer type.
                return kubeconfigRequest, nil
```

Reply:
✅  (I also omitted the first two comments as they did not add much value)